### ISSUE TITLE
Simplify/robustify LogSoftmax backward calc

### ISSUE THREAD

**Issue by yv** [clab/dynet#218]
The gradient value for logsoftmax should be something like "grad - sum(grad)*softmax(x)" (which Theano and TensorFlow use).
```
z.t<0>().device(*dev.edevice) = fx.t<1>().binaryExpr(dEdf.t<1>(), FWeightedError()).sum();
Eigen::array<int, 1> bcast; bcast[0] = fx.d.rows();
dEdxi.t<1>().device(*dev.edevice) += fx.t<1>().exp() * -z.t<1>().broadcast(bcast) + dEdf.t<1>();
```
or, in cnn1
`float off_diag_sum = -(*fx).binaryExpr(*dEdf, FWeightedError()).sum();`
where FWeightedError calculates `expf(t) * d / expf(t)`, which is `d` in most cases and **nan** in cases where t is large and negative, (which means the nan will propagate and clip_gradients will abort the program).

Simplifying this to the moral equivalent of
`float off_diag_sum = -(*dEdf).sum();`
would remove the occurrence of nans in some cases.
